### PR TITLE
Add syntax highlight to datafusion-cli

### PR DESCRIPTION
**Pull request by trungda**
## Which issue does this PR close?
https://github.com/apache/arrow-datafusion/issues/8701

## Rationale for this change
More user-friendly and readability when using `datafusion-cli`.

## What changes are included in this PR?
Add SQL syntax highlighting capability for `datafusion-cli`: `SyntaxHighlighter`.

## Are these changes tested?
1. Added unit tests to validate the ANSI string generated;
2. Manually tested.
Screenshot from 2024-01-19 14-29-03

## Are there any user-facing changes?
Users of datafusion-cli.

### CODE REVIEW

**Comment by trungda** [comment]
The current implementation is not the best. If the input string is invalid, e.g., unterminated: `select * from tab_1 where col_1 = '`, the whole string will be _un_highlighted. I think a better approach is to make change to `sqlparser` crate to expose a token iterator just so we can keep highlighting as we scan through the original string, and only stop when we encounter a failure. Something along this line: https://github.com/sqlparser-rs/sqlparser-rs/blob/main/src/tokenizer.rs#L555-L558.

I am happy to go down that path too. Let me know :)

**Review by comphead — COMMENTED** [review]
I really love this one, thanks @trungda 
I checked it in the codespace, it has some bugs mostly on when you type a key word and then press left arrow.

**Comment by comphead** [comment]
image

**Comment by comphead** [comment]
image

**Comment by trungda** [comment]
> image

Thanks for the feedback @comphead . What is your terminal setup? I am not able to repro yet. The current approach is to take the whole input string and tokenize the input as a whole, the arrow keys shouldn't matter. 

Screenshot from 2024-01-20 13-38-32

**Comment by Weijun-H** [comment]
Awesome feature! Thanks, @trungda 👍 . I am wondering if could make the default color lighter because the current version is hard to track. Maybe you could consider the DuckDB setting.
image

**Comment by alamb** [comment]
> I am happy to go down that path too. Let me know :)

That would be great -- thank you @trungda  (this is somewhat related); https://github.com/sqlparser-rs/sqlparser-rs/pull/1094

> On that note, can we add datafusion-cli to members attribute in [workspace] just so it will be recognized by the rust-analyzer plugin in IDEs?

The reason it is not part of the workspace is that as `datafusion-cli` we want to release a binary and thus have the versions locked (via a checked in `Cargo.lock`) However, for the `datafusion` crates we release them as a library where having Cargo.lock is not the recommended.

This is not clear I will add it to the README

File: datafusion-cli/src/highlighter.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        };
+        Self { dialect }
+    }
+}
+
+impl Highlighter for SyntaxHighlighter {
+    fn highlight<'l>(&self, line: &'l str, _: usize) -> Cow<'l, str> {
+        let mut out_line = String::new();
+
+        // `with_unescape(false)` since we want to rebuild the original string.
```
Comment (alamb) [inline datafusion-cli/src/highlighter.rs:51]: 💯

File: datafusion-cli/src/highlighter.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline datafusion-cli/src/highlighter.rs:1]: For some reason the RAT check fails for this file: https://github.com/apache/arrow-datafusion/actions/runs/7592786554/job/20702838429?pr=8918

I think that can typically be fixed by just copy/pasting the license text / comment from another file

**Review by alamb — APPROVED** [review]
Thank you @trungda  -- I tried this out and it was great. 

Once CI passes I think we could merge it. 

In terms of changing the default colors / improving the highlighting, it would be great to improve this PR but also I think we can iterate in the future as well

Another feature that would be neat would be a CLI flag to enable/disable the highlighting so people who find the color scheme not to their liking could turn it off

File: datafusion-cli/src/highlighter.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apndache.org/licenses/LICENSE-2.0
```
Comment (Jefffrey) [inline datafusion-cli/src/highlighter.rs:9]:
```suggestion
//   http://www.apache.org/licenses/LICENSE-2.0
```

typo in license

**Comment by trungda** [comment]
Thanks @alamb ! I've updated the license + changed the color to be lighter as suggested. 

> That would be great -- thank you @trungda (this is somewhat related); https://github.com/sqlparser-rs/sqlparser-rs/pull/1094

This is great! I'll look into that. It'd be easier if somehow we know the position in the original string where the tokernization fails just so we can just copy from that position to the highlighted string.

> Another feature that would be neat would be a CLI flag to enable/disable the highlighting so people who find the color scheme not to their liking could turn it off

Sounds great! Let me look into that after this PR.

File: datafusion-cli/src/highlighter.rs (hunk 2 above)
Comment (trungda) [reply datafusion-cli/src/highlighter.rs:1]: Thanks! There was a typo in the license as pointed about by @Jefffrey . Let's see if it's green now :)

**Review by Weijun-H — APPROVED** [review]
LGTM! Thanks @trungda

**Pull request merged**